docs/custom-plugins/custom-rules.md

```diff
@@ -0,0 +1,117 @@
+      if (operation.operationId === 'test') {
+        ctx.report({
+          message: `operationId must be not "test"`,
+          location: ctx.location.child('operationId'),
+        });
+      }
+    },
+  };
+}
+
+```
+
+The `ctx` object here holds all the context, which can be used to give more situation-aware functionality to the rules you build. This is one of the main use cases for custom rules. The `report()` method is used to give information to return to the user if the node being visited doesn't comply with the rule. You can read the [context](#the-context-object) and [location](#location-object) sections for more information.
+
+Adding this as part of a plugin requires you to add it to the `rules` part of the plugin object, under the relevant document type. The example rule here is intended to be used with OpenAPI, so the plugin code in `plugins/my-rules.js` is as follows:
+
+```js
+const OpIdNotTest = require('./rules/opid-not-test.js');
+
+module.exports = {
+  id: 'my-rules',
+  rules: {
+    oas3: {
+      'opid-not-test': OperationIdNotTest,
+    }
+  },
+};
+```
+
+To use the example rule, add the following to your `redocly.yaml` configuration file:
+
+```yaml
+plugins:
+  - 'plugins/my-rules.js'
+
+rules:
+  my-rules/opid-not-test: warn
+```
+
+Validate your OpenAPI document with `redocly lint openapi.yaml`. With this rule enabled as shown, any `operationId` fields that are "test" cause the validation step to emit a warning. You can use the example here as a basis for building your own rules, and many rules can be included in a single plugin.
+
+## Object references
+
+### The context object
+
+The context object contains additional functionality that is helpful for rules to do their jobs. As the name implies, the context object contains information that is relevant to the context of the rule. The context object has the following properties:
+
+- `location` - Current location in the source document. See [Location Object](#location-object).
+- `parentLocations` - Mapping of the parent node to its location (only for nested visitors).
+- `type` - Information about the current type from the type tree.
+- `parent` - Parent object or array.
+- `key` - Key in the parent object or array.
+- `oasVersion` - Specific OAS minor version of the current document (can be `oas2`, `oas3` or `oas3_1`).
+
+The context object also offers some additional functionality to resolve references and to return information about a problem to the user. The methods available are as follows:
+
+- `resolve(node)` - Synchronously dereferences `$ref` node to its value. Works only with `$refs` from the original document. If you need to resolve a reference from another source, you can use the optional second parameter: `resolve(node, from: string)`.
+- `report(descriptor)` - Reports a problem in the definition and returns information to the user. See [Report rule context](#report-rule-context) for more information.
+
+<a id="context-report"></a>
```

Comment:
suggestion: Remove this anchor because this is a new page and there are no pre-existing links.
```suggestion
```